### PR TITLE
implement many copy flags

### PR DESCRIPTION
**Pull request by millerjs**
I know there are at least two other open PRs for `cp` so I figured I throw this one up that I started some time ago in case it's useful.

I'd like to finish this up, but just haven't gotten around to it.

cc @Matt8898 @jerenept  @Arcterus

- [x] backup
- [x] force
- [x] interactive
- [x] link
- [x] no-clobber
- [x] no-target-directory
- [x] paths
- [x] recursive
- [x] remove-destination
- [x] suffix
- [x] symbolic-link
- [x] target-directory
- [x] verbose
- [x] version
- [ ] archive
- [ ] attributes-only
- [ ] copy-contents
- [ ] no-dereference-preserve-linkgs
- [ ] dereference
- [ ] no-dereference
- [ ] preserve-default-attributes
- [ ] preserve
- [ ] no-preserve
- [ ] parents
- [ ] reflink
- [ ] sparse
- [ ] strip-trailing-slashes
- [ ] update
- [ ] one-file-system
- [ ] context
- [ ] cli-symbolic-links

### CODE REVIEW

**Comment by millerjs** [comment]
In the context of other `cp` PRs (e.g. https://github.com/uutils/coreutils/pull/1052), I'm looking for input on whether this is worth merging (granted the only test failing is, windows --force, is fixed) or whether it will be ignored 

cc @mpkh?

**Comment by Arcterus** [comment]
I think this is definitely worth merging!  I would prefer that you clean up the commits a bit though.

File: src/cp/Cargo.toml [hunk 1]
```diff
@@ -8,10 +8,12 @@ name = "uu_cp"
 path = "cp.rs"
 
 [dependencies]
+walkdir = "1"
+quick-error = "1.1.0"
 getopts = "*"
 libc = "*"
 uucore = { path="../uucore" }
-walkdir = "*"
+clap = "*"
```
Comment (Arcterus) [inline src/cp/Cargo.toml]: Can you change this to a specific version?  I think we should probably start specifying versions for all of the utils' dependencies...

File: src/cp/cp.rs [hunk 2]
```diff
@@ -4,245 +4,883 @@
...
+    let matches = App::new(executable!())
+        .version(VERSION)
+        .about(&about[..])
+        .arg(Arg::with_name(OPT_TARGET_DIRECTORY)
+             .short("t")
+             .conflicts_with(OPT_NO_TARGET_DIRECTORY)
+             .long(OPT_TARGET_DIRECTORY)
+             .value_name(OPT_TARGET_DIRECTORY)
+             .takes_value(true)
+             .conflicts_with(OPT_NO_TARGET_DIRECTORY)
```
Comment (Arcterus) [inline src/cp/cp.rs]: You have `conflicts_with` twice.

**Comment by millerjs** [comment]
definitely! Happy to address those requests, just wanted to check if it was worth continuing.

I'll fix tests, come to a stoping point and clean up the code. Thanks!

**Comment by millerjs** [comment]
@Arcterus I've marked `--force` and `--remove-destination` tests skipped for now on Windows because I'm not sure what's failing when attempting to remove readonly files during the opaque debug cycle of pushing to CI.  Otherwise it should be good to go

**Comment by Arcterus** [comment]
Is it just the tests that are failing, or do `--force` and `--remove-destination` not work on Windows?

**Comment by millerjs** [comment]
I believe it's `fs::remove_file(dest)` on a readonly file that is failing on Windows, so `--force` won't work (it is marked unimplemented on windows) and `--remove-destination` will only work if the file is writeable.  I don't have a Windows machine however, and debugging it has eluded me while pushing to CI. I tried unsetting the readonly property before removing the file on Windows but that did not seem to work.

**Comment by Arcterus** [comment]
Alright, well, for now can you add something to the help message for `--remove-destination` mentioning that it only works on writable files for Windows?  I'll merge this after you do that. :)

**Comment by millerjs** [comment]
sounds good to me, will add and squash again

**Comment by Arcterus** [comment]
Thanks :+1:

**Pull request merged**